applications/FluidDynamicsApplication/custom_elements/embedded_navier_stokes.h

```diff
@@ -658,240 +552,201 @@ class EmbeddedNavierStokes : public NavierStokes<TDim, TNumNodes>
             {
-                //~ rRightHandSideVector(i*BlockSize+row) -= auxRightHandSideVector(i*TDim+row); // TODO: Check these signs
-                rRightHandSideVector(i*BlockSize+row) += auxRightHandSideVector(i*TDim+row);
+                diag_max = fabs(rLeftHandSideMatrix(i,i)); // Maximum diagonal value (associated to velocity)
             }
         }
 
+        const double K = 100.0;
+        const double denominator = std::max(0.0001*rData.h*rData.h, rSplittingData.intersection_area);
+        const double pen_coef = K*diag_max/denominator;
+
+        return pen_coef;
     }
 
 
+    /**
+    * This functions adds the penalty extra term level set contribution.
+    * @param rLeftHandSideMatrix: reference to the LHS matrix
+    * @param rRightHandSideVector: reference to the RHS vector
+    * @param rData: reference to element data structure
+    * @param rSplittingData: reference to the intersection data structure
+    */
     void AddBoundaryConditionPenaltyContribution(MatrixType& rLeftHandSideMatrix,
                                                  VectorType& rRightHandSideVector,
-                                                 const ElementDataType& data,
-                                                 const unsigned int& cut_points,
-                                                 const VectorType& cut_areas,
-                                                 const MatrixType& Ncontainer_cut)
+                                                 const ElementDataType& rData,
+                                                 const ElementSplittingDataStruct& rSplittingData)
     {
-        constexpr unsigned int BlockSize = TDim+1;                 // Block size
-        constexpr unsigned int MatrixSize = TNumNodes*BlockSize;   // Matrix size
+        constexpr unsigned int BlockSize = TDim+1;
+        constexpr unsigned int MatrixSize = TNumNodes*BlockSize;
 
         // Obtain the previous iteration velocity solution
         array_1d<double, MatrixSize> prev_sol = ZeroVector(MatrixSize);
-
-        for (unsigned int i=0; i<TNumNodes; i++)
-        {
-            for (unsigned int comp=0; comp<TDim; comp++)
-            {
-                prev_sol(i*BlockSize+comp) = data.v(i,comp);
-            }
-        }
+        GetPreviousSolutionVector(rData, prev_sol);
 
         MatrixType P_gamma(TNumNodes, TNumNodes);    // Penalty matrix
         noalias(P_gamma) = ZeroMatrix(TNumNodes, TNumNodes);
 
-        VectorType aux_cut(TNumNodes);
-
-        double intersection_area = 0.0;
-
-        for (unsigned int icut = 0; icut<cut_points; icut++)
+        for (unsigned int icut = 0; icut<rSplittingData.ncutpoints; icut++)
         {
-            double weight = cut_areas(icut);
-            intersection_area += weight;
-
-            aux_cut = row(Ncontainer_cut, icut);
+            const double weight = rSplittingData.cut_edge_areas(icut);
+            const VectorType aux_cut = row(rSplittingData.N_container_cut, icut);
```

Comment:
i believe you know the size at compile type of aux_cut. if so try if possible to use a stack array instead